### ISSUE TITLE
exp: pilot contacts-v1 `1.5B` tuning sweep

### ISSUE THREAD

**Issue by eric-czech** [Open-Athena/MarinFold#75]
Sweep LR, weight decay and epochs at 1.5B params on 4.6B token `contacts-v1` dataset [[1](https://github.com/Open-Athena/MarinFold/issues/61)] (loosely following [[2](https://arxiv.org/pdf/2509.14786)]), then compare to https://github.com/Open-Athena/MarinFold/pull/70.  If tuning is productive, repeat across model scales.

## Details

- **Branch:** `eac/plm-exp75`
- **Data:** train cache `tokenized/contacts-v1-663ba6` (~4.7B tok), val cache `tokenized/contacts-v1-val-92827b` (41,954 docs) — from [`hf://buckets/open-athena/MarinFold/.../contacts_v1`](https://huggingface.co/buckets/open-athena/MarinFold/tree/data/document_structures/contacts_v1) (`{train,val}`)
- **Packing:** `pack=True`, prefix-only; no cross-document attn
- **Tokenization:** `timodonnell/contacts-v1-tokenizer@5d68a24a899f` (2845 vocab)
- **Loss:** unmasked (contacts-v1 has no `<distance>` statements)
- **Eval:** full held-out val split, no downsampling; 2 evals/epoch
- **Checkpoint:** temp every 10 min; 1 permanent
- **Arch:** Qwen3; `seq_len=8192`; ~1.47B params (exp44 dims → `hidden_dim=2048`, `intermediate_dim=8192`, `num_layers=24`, `num_heads=32`, `num_kv_heads=8`, `rope=Llama3RotaryEmbeddingsConfig()`)
- **Optimizer/LR:** AdamW, cosine schedule, 10% warmup; LR / weight decay / epochs swept
- **Batch:** global 128, `seq_len=8192` (~1.05M tok/step; 1 epoch ≈ 4,490 steps)
- **Shuffle/Seed:** full Feistel permutation over train, `data_seed=0` (shards are round-descending, highest-pLDDT last)

## Results

See https://github.com/Open-Athena/MarinFold/issues/154 for checkpoints, wandb runs, code and analysis ([here](https://github.com/Open-Athena/MarinFold/issues/154#issuecomment-5192141278)).

**Comment by eric-czech** [comment]
## Results

Sweep complete — 63 finished runs over LR × weight-decay × {1, 2, 4, 8} epochs (8 crashed, none retried). Best config is **2.7566** `eval/contacts-v1-val/loss` (lr=1e-3, wd=0.2, 8 epochs), beating the 1.5B unmasked baseline from #70 (**2.9800**); 31 configs beat it. Epoch count is the dominant lever — LR/WD tuning moves loss ~0.02–0.05 within an epoch rung, while 1→8 epochs moves it ~0.29.

<img width="900" alt="loss vs completion time" src="https://gist.githubusercontent.com/eric-czech/109309283a67061bd7f0fb42346ff701/raw/loss_vs_completion.png" />

Final loss for every finished run against wall-clock completion time (so this doubles as sweep progress); the dashed line is the running best-so-far and the solid purple line the 1.5B unmasked baseline (#70).

---

<p align="center">
<img width="500" alt="best loss vs epoch count" src="https://gist.githubusercontent.com/eric-czech/109309283a67061bd7f0fb42346ff701/raw/best_loss_vs_epochs.png" />
</p>

Best run at each epoch count over the entire LR/WD grid (x is log2) — the baseline is first cleared at 2 epochs, and the largest single jump is 4→8.

---

<img width="900" alt="loss curves by epoch" src="https://gist.githubusercontent.com/eric-czech/109309283a67061bd7f0fb42346ff701/raw/loss_curves_by_epoch.png" />

Per-run eval loss vs cumulative tokens; axes are independent per facet and points above 3.4 are clipped to up-triangles, so the four panels aren't on a common scale.

---

<img width="900" alt="contact metrics vs loss" src="https://gist.githubusercontent.com/eric-czech/109309283a67061bd7f0fb42346ff701/raw/contact_metrics_vs_loss.png" />

Downstream contact AUC and R-precision (all/long) for each epoch's best checkpoint vs its loss (single realization, scored via the #89 harness; loss axis reversed) — contact skill appears abruptly at E8 even though the loss gains look smooth.

---

<details>
<summary><b>LR / weight-decay heatmap</b> (final loss per epoch count)</summary>

<img width="900" alt="LR/WD heatmap by epoch" src="https://gist.githubusercontent.com/eric-czech/109309283a67061bd7f0fb42346ff701/raw/lr_wd_heatmap.png" />

Final loss over the LR×WD grid per epoch count with independent color scales; blank cells were not run at that epoch count and each panel outlines its best cell.

</details>

---

<details><summary><b>Compute usage summary</b> (FLOPs and time)</summary>

From 63 `finished` runs matching `prot-exp75-.*-v1`; `runtime` is "Runtime" in UI (start to last reported activity):

```
By epochs

┌────────┬────────┬────────────┬──────────────────┬─────────────┐
│ epochs │ # runs │  Σ PFLOP   │    Σ runtime     │ runtime (h) │
├────────┼────────┼────────────┼──────────────────┼─────────────┤
│ 1      │     15 │    959,727 │  14d 08h 14m 31s │      344.24 │
├────────┼────────┼────────────┼──────────────────┼─────────────┤
│ 2      │     25 │  3,199,092 │  29d 13h 51m 50s │      709.86 │
├────────┼────────┼────────────┼──────────────────┼─────────────┤
│ 4      │     11 │  2,815,201 │  43d 07h 24m 45s │    1,039.41 │
├────────┼────────┼────────────┼──────────────────┼─────────────┤
│ 8      │     12 │  6,142,256 │  46d 15h 30m 21s │    1,119.51 │
├────────┼────────┼────────────┼──────────────────┼─────────────┤
│ Total  │     63 │ 13,116,275 │ 133d 21h 01m 27s │    3,213.02 │
└────────┴────────┴────────────┴──────────────────┴─────────────┘

By TPU slice type (tpu= tag)   

┌────────┬────────┬────────────┬──────────────────┬─────────────┐
│ slice  │ # runs │  Σ PFLOP   │    Σ runtime     │ runtime (h) │
├────────┼────────┼────────────┼──────────────────┼─────────────┤
│ v5p-8  │      4 │    319,909 │   4d 19h 37m 11s │      115.62 │
├────────┼────────┼────────────┼──────────────────┼─────────────┤
│ v5p-16 │      7 │  2,815,201 │  29d 16h 28m 28s │      712.47 │
├────────┼────────┼────────────┼──────────────────┼─────────────┤
│ v5p-32 │     13 │  3,007,146 │  18d 23h 20m 33s │      455.34 │
├────────┼────────┼────────────┼──────────────────┼─────────────┤
│ v5p-64 │      5 │  1,663,528 │   9d 15h 46m 49s │      231.78 │
├────────┼────────┼────────────┼──────────────────┼─────────────┤
│ v6e-4  │      3 │    191,945 │   4d 02h 37m 17s │       98.62 │
├────────┼────────┼────────────┼──────────────────┼─────────────┤
│ v6e-8  │     26 │  4,094,837 │  59d 05h 16m 47s │    1,421.28 │
├────────┼────────┼────────────┼──────────────────┼─────────────┤
│ v6e-16 │      4 │    511,855 │   3d 12h 10m 10s │       84.17 │
├────────┼────────┼────────────┼──────────────────┼─────────────┤
│ v6e-32 │      1 │    511,855 │   3d 21h 44m 12s │       93.74 │
├────────┼────────┼────────────┼──────────────────┼─────────────┤
│ Total  │     63 │ 13,116,275 │ 133d 21h 01m 27s │    3,213.02 │
└────────┴────────┴────────────┴──────────────────┴─────────────┘

By TPU family - total-hours

┌────────┬────────┬────────────┬──────────────────┬─────────────┐
│ family │ # runs │  Σ PFLOP   │    Σ runtime     │ runtime (h) │
├────────┼────────┼────────────┼──────────────────┼─────────────┤
│ v5p    │     29 │  7,805,783 │  63d 03h 13m 01s │    1,515.22 │
├────────┼────────┼────────────┼──────────────────┼─────────────┤
│ v6e    │     34 │  5,310,492 │  70d 17h 48m 26s │    1,697.81 │
├────────┼────────┼────────────┼──────────────────┼─────────────┤
│ Total  │     63 │ 13,116,275 │ 133d 21h 01m 27s │    3,213.02 │
└────────┴────────┴────────────┴──────────────────┴─────────────┘


By TPU family — chip-hours

┌────────┬────────┬────────────┬────────────┐
│ family │ # runs │  Σ PFLOP   │ chip-hours │
├────────┼────────┼────────────┼────────────┤
│ v5p    │     29 │  7,805,783 │  20,864.72 │
├────────┼────────┼────────────┼────────────┤
│ v6e    │     34 │  5,310,492 │  16,111.01 │
├────────┼────────┼────────────┼────────────┤
│ Total  │     63 │ 13,116,275 │  36,975.73 │
└────────┴────────┴────────────┴────────────┘

```

</details>

---

<details><summary><b>Checkpoint paths</b> (GCS)</summary>

Checkpoints for best models by epoch count ([slack](https://openathena.slack.com/archives/C0AB3K2G6S0/p1782226874577869?thread_ts=1782160108.345019&cid=C0AB3K2G6S0)):

```
ep	cell	    loss	checkpoint path
E1	7e-4/wd0.05	3.046	gs://marin-us-east5/checkpoints/prot-exp75-cv1-1_5b-e1-lr7e-4-wd0p05-v1-69de44/
E2	7e-4/wd0.8	2.942	gs://marin-us-east5/checkpoints/prot-exp75-cv1-1_5b-e2-lr7e-4-wd0p8-v1-d4dfdb/
E4	1e-3/wd0.05	2.924	gs://marin-us-east5/checkpoints/prot-exp75-cv1-1_5b-e4-lr1e-3-wd0p05-v1-dcd938/
E8	1e-3/wd0.2	2.757	gs://marin-us-east5/checkpoints/prot-exp75-cv1-1_5b-e8-lr1e-3-wd0p2-v1-bc3084/
```

</details>

---

- raw results: [exp75_results.zip](https://github.com/user-attachments/files/29807071/exp75_results.zip)
- wandb: [runs](https://wandb.ai/eric-czech/marin/groups/exp75-contacts-v1-tune)
- sweep: [`experiments/protein/exp75_sweep.py`](https://github.com/marin-community/marin/blob/be811f88706bc6dbb1881ff260eb6772a474803d/experiments/protein/exp75_sweep.py)
- analysis: [`experiments/protein/exp75_analysis.py`](https://github.com/marin-community/marin/blob/be811f88706bc6dbb1881ff260eb6772a474803d/experiments/protein/exp75_analysis.py)
- checkpoints: [`open-athena/marinfold-exp75`](https://huggingface.co/open-athena/marinfold-exp75)
